vectorlm/utils/save_utils.py

```diff
@@ -187,51 +159,67 @@ def save_optimizer(
         rank: The worker's rank.
 
     """
-    opt_name = f"optimizer_rank{rank}.bin"
-    output_optimizer_file = os.path.join(output_dir, opt_name)
-    opt_cfg = LocalOptimStateDictConfig(offload_to_cpu=True)
+    os.makedirs(output_dir, exist_ok=True)
+    opt_cfg = ShardedOptimStateDictConfig(offload_to_cpu=True)
     with FSDP.state_dict_type(
         model,
-        StateDictType.LOCAL_STATE_DICT,
+        StateDictType.SHARDED_STATE_DICT,
         optim_state_dict_config=opt_cfg,
     ):
-        opt_state = FSDP.optim_state_dict(model, optimizer)
-        print(f"Saving optimizer state to {output_optimizer_file}")
-        torch.save(opt_state, output_optimizer_file)
-        print(f"Optimizer state saved to {output_optimizer_file}")
+        state_dict = model.state_dict()
+        opt_state = FSDP.sharded_optim_state_dict(model, optimizer)
+        full_state = {"model_state": state_dict, "optim_state": opt_state}
+    writer = FileSystemWriter(output_dir, single_file_per_rank=True)
+    if _should_save(rank, model.sharding_strategy):
+        if rank == 0:
+            print(f"Saving states to {output_dir}")
+        save(
+            state_dict=full_state,
+            storage_writer=writer,
+            process_group=model.process_group,
+            planner=DefaultSavePlanner(),
+        )
+        if rank == 0:
+            print(f"States saved to {output_dir}")
 
 
-def load_optimizer(
+def load_model_optimizer(
```

Comment:
Same comment as above. Not sure if this is clear that it loads both the model and optimizer.